modules/runtime/nuxeo-runtime/src/main/java/org/nuxeo/runtime/RuntimeMessage.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * (C) Copyright 2017-2020 Nuxeo (http://nuxeo.com/) and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Kevin Leturc <kleturc@nuxeo.com>
+ *     Anahide Tchertchian
+ */
+package org.nuxeo.runtime;
+
+import java.util.logging.Level;
+
+/**
+ * @since 11.3
+ */
+public class RuntimeMessage {
+
+    protected final ComponentManagerStep step;
+
+    protected final Level level;
+
+    protected final String message;
+
+    protected final String source;
+
+    public RuntimeMessage(ComponentManagerStep step, Level level, String message) {
+        this(step, level, message, null);
+    }
+
+    public RuntimeMessage(ComponentManagerStep step, Level level, String message, String source) {
+        super();
+        this.step = step;
+        this.level = level;
+        this.message = message;
+        this.source = source;
+    }
+
+    public ComponentManagerStep getStep() {
+        return step;
+    }
+
+    public Level getLevel() {
+        return level;
+    }
+
+    public String getMessage() {
+        return message;
+    }
+
+    public String getSource() {
+        return source;
+    }
+
+    protected enum ComponentManagerStep {
+
+        /**
+         * Pseudo state when listener has not been called yet.
```

Comment:
Which listener?

Reply:
RuntimeMessageHandler is a runtime listener. Maybe we don't need this info anymore, wdyt @kevinleturc ?

Reply:
We can get rid of the listener aspect there indeed.

The new state `INITIALIZING` represents the ComponentManager state before it starts its activation and it is only reachable just after the ComponentManager instantiation right?

We can reword the comment in this way 👍 .

Reply:
If we don't need this to be a listener, i'll get rid of the whole "step" information anyway. So which is it? Do you think we'll need to know at which step the message has been added?
My understanding is that this info was added to handle dupe messages in case of hotreload, but the strategy is not working anyway, that's why i added the notion of source instead.

Reply:
>If we don't need this to be a listener, i'll get rid of the whole "step" information anyway. So which is it? Do you think we'll need to know at which step the message has been added?

We don't really need to know precisely the step, in fact we need to know if it happens during ACTIVATING + STARTING, or before RUNNING in order to not clean them before displaying them.

The source approach is interesting but it requires to manipulate the stack of messages outside of the handler which is not great to handle all usages (runtime/hotreload/test) and requires additional API for this. In addition, see one of my comments, it requires to put `ComponentManagerImpl` as a source for all `Component`. Is it really source? We won't be able to use this new field to know which component reported the message, or am I missing something?

Reply:
I cannot finely know which component provided the registration without changing the DescriptorRegistry api, that's why i limited changes here by cleaning up all messages from the ComponentManager source, even if i'm not happy with these changes. We're not relying on the step at all anymore: calls to #applyStash is the current criterion i'm using.

I chose to go this way for now because at least we're not hiding important problemes at startup anymore: it fixes the duplicate component name detection at startup.

Now if we want sthg finer, i'll postpone all this work until i get a go to fix the registration in general (as current code is nonsense): this will take time and bring more questions/issues, so this may not happen any time soon.

Reply:
relying on RUNNING state to clear messages fixed it --> cleaned up the step on messages, but kept the listener logics to be able to clear all messages when *previous* step was RUNNING